src/darwin/Framework/CHIP/MTRSetupPayload.mm

```diff
@@ -19,251 +19,503 @@
+    VerifyOrReturnValue(self.type != MTROptionalQRCodeInfoTypeUnknown, nil);
+    return self;
 }
 
-+ (chip::Optional<chip::RendezvousInformationFlags>)convertDiscoveryCapabilities:(MTRDiscoveryCapabilities)value
+- (CHIP_ERROR)addAsVendorElementTo:(chip::SetupPayload &)payload
 {
-    if (value == MTRDiscoveryCapabilitiesUnknown) {
-        return chip::NullOptional;
+    switch (_info.type) {
+    case chip::optionalQRCodeInfoTypeString:
+        return payload.addOptionalVendorData(_info.tag, _info.data);
+    case chip::optionalQRCodeInfoTypeInt32:
+        return payload.addOptionalVendorData(_info.tag, _info.int32);
+    default:
+        return CHIP_ERROR_INCORRECT_STATE;
     }
+}
 
-    chip::RendezvousInformationFlags flags;
-    if (value & MTRDiscoveryCapabilitiesBLE) {
-        flags.Set(chip::RendezvousInformationFlag::kBLE);
-    }
+- (MTROptionalQRCodeInfoType)type
+{
+    // The way OptionalQRCodeInfo uses what are effectively C type identifiers (uint32 etc)
+    // rather than TLV types is not ideal. If we add support for additional integer types
+    // we should consider replacing MTROptionalQRCodeInfoTypeInt32 with
+    // MTROptionalQRCodeInfoTypeInteger and hiding the low-level C representation.
+    switch (_info.type) {
+    case chip::optionalQRCodeInfoTypeString:
+        return MTROptionalQRCodeInfoTypeString;
+    case chip::optionalQRCodeInfoTypeInt32:
+        return MTROptionalQRCodeInfoTypeInt32;
+    // No 'default:' so we get a warning if new types are added.
+    // OptionalQRCodeInfo does not support these types
+    case chip::optionalQRCodeInfoTypeInt64:
+    case chip::optionalQRCodeInfoTypeUInt32:
+    case chip::optionalQRCodeInfoTypeUInt64:
+    // We should never see the unknown type
+    case chip::optionalQRCodeInfoTypeUnknown:
+        /* fall through */;
+    }
+    return MTROptionalQRCodeInfoTypeUnknown;
+}
+
+- (uint8_t)tagNumber
+{
+    return _info.tag;
+}
+
+- (NSNumber *)integerValue
+{
+    VerifyOrReturnValue(_info.type == chip::optionalQRCodeInfoTypeInt32, nil);
+    return @(_info.int32);
+}
+
+- (NSString *)stringValue
+{
+    VerifyOrReturnValue(_info.type == chip::optionalQRCodeInfoTypeString, nil);
+    return [NSString stringWithUTF8String:_info.data.c_str()];
+}
+
+- (id)copyWithZone:(NSZone *)zone
+{
+    return self; // immutable
+}
+
+- (NSUInteger)hash
+{
+    return _info.type << 8 | _info.tag;
+}
+
+- (BOOL)isEqual:(id)object
+{
+    VerifyOrReturnValue([object class] == [self class], NO);
+    MTROptionalQRCodeInfo * other = object;
+    VerifyOrReturnValue(_info.tag == other->_info.tag, NO);
+    VerifyOrReturnValue(_info.type == other->_info.type, NO);
+    VerifyOrReturnValue(_info.int32 == other->_info.int32, NO);
+    VerifyOrReturnValue(_info.data == other->_info.data, NO);
+    return YES;
+}
+
+- (NSString *)description
+{
+    return [NSString stringWithFormat:@"0x%02x=%@", self.tagNumber, self.integerValue ?: self.stringValue];
```

Comment:
Any reason we are not including the class name here?

Reply:
I figured it would give nicer output if you print eg the whole list, and the class name doesn’t add that much here. (For comparison, NSNumber and NSString also don’t include the class name in the description)